tests/components/generic/test_camera.py

```diff
@@ -63,10 +66,86 @@ async def test_fetching_url(
     body = await resp.read()
     assert body == fakeimgbytes_png
 
+    # sleep .1 seconds to make cached image expire
+    await asyncio.sleep(0.1)
+
     resp = await client.get("/api/camera_proxy/camera.config_test")
     assert respx.calls.call_count == 2
 
 
+@respx.mock
+async def test_image_caching(
+    hass: HomeAssistant, hass_client: ClientSessionGenerator, fakeimgbytes_png
+) -> None:
+    """Test that the image is cached and not fetched more often than the framerate indicates."""
+    respx.get("http://example.com").respond(stream=fakeimgbytes_png)
+
+    framerate = 5
+    await async_setup_component(
+        hass,
+        "camera",
+        {
+            "camera": {
+                "name": "config_test",
+                "platform": "generic",
+                "still_image_url": "http://example.com",
+                "username": "user",
+                "password": "pass",
+                "authentication": "basic",
+                "framerate": framerate,
+            }
+        },
+    )
+    await hass.async_block_till_done()
+
+    client = await hass_client()
+
+    initial_time = datetime.now()
+    with freeze_time(initial_time) as frozen_datetime:
+        resp = await client.get("/api/camera_proxy/camera.config_test")
+        assert resp.status == HTTPStatus.OK
+        body = await resp.read()
+        assert body == fakeimgbytes_png
+
+        resp = await client.get("/api/camera_proxy/camera.config_test")
+        assert resp.status == HTTPStatus.OK
+        body = await resp.read()
+        assert body == fakeimgbytes_png
+
+        # time is frozen, image should have come from cache
+        assert respx.calls.call_count == 1
+
+        # advance time by 100ms
+        frozen_datetime.move_to(initial_time + timedelta(0, 0.1))
+
+        resp = await client.get("/api/camera_proxy/camera.config_test")
+        assert resp.status == HTTPStatus.OK
+        body = await resp.read()
+        assert body == fakeimgbytes_png
+
+        # Only 100ms have passed, image should still have come from cache
+        assert respx.calls.call_count == 1
+
+        # advance time by another 100ms
+        frozen_datetime.move_to(initial_time + timedelta(0, 0.2))
```

Comment:
I think this could be a knife edge condition. Depending on how much time elapsed from when you took the `now()` reference, to when the freeze became effective, you may have less than 200ms between the first image being fetched and the 200ms marker.

To avoid this, you could just use a 300ms marker.

You might want to also consider adding `freezer: FrozenDateTimeFactory` to the parameters of the test function. I believe that freezes time for you at the start, then you can do `freezer.tick(interval)` without having to keep track of the start time.  This is a common pattern in the tests of other components (but both methods are used and both work!)

Reply:
changed to `FrozenDateTimeFactory` and updated the interval to 150ms (and 300ms total) to avoid that condition.

Reply:
btw, here's a related PR that fixes the interval calculation between image fetches:
https://github.com/home-assistant/core/pull/105820